darshan-util/darshan-logutils-accumulator.c

```diff
@@ -0,0 +1,321 @@
+    double *rank_cumul_rw_only_time;
+    double *rank_cumul_md_only_time;
+};
+
+int darshan_accumulator_create(darshan_module_id id,
+                               int64_t job_nprocs,
+                               darshan_accumulator*   new_accumulator)
+{
+    *new_accumulator = NULL;
+
+    if(id >= DARSHAN_KNOWN_MODULE_COUNT)
+        return(-1);
+
+    if(!mod_logutils[id]->log_agg_records ||
+       !mod_logutils[id]->log_sizeof_record ||
+       !mod_logutils[id]->log_record_metrics) {
+        /* this module doesn't support this operation */
+        return(-1);
+    }
+
+    *new_accumulator = calloc(1, sizeof(struct darshan_accumulator_st));
+    if(!(*new_accumulator))
+        return(-1);
+
+    (*new_accumulator)->module_id = id;
+    (*new_accumulator)->job_nprocs = job_nprocs;
+    (*new_accumulator)->agg_record = calloc(1, DEF_MOD_BUF_SIZE);
+    if(!(*new_accumulator)->agg_record) {
+        free(*new_accumulator);
+        *new_accumulator = NULL;
+        return(-1);
+    }
+    /* 3 arrays handled in one malloc */
+    (*new_accumulator)->rank_cumul_io_total_time = calloc(job_nprocs*3, sizeof(double));
+    if(!(*new_accumulator)->rank_cumul_io_total_time) {
+        free((*new_accumulator)->agg_record);
+        free(*new_accumulator);
+        *new_accumulator = NULL;
+        return(-1);
+    }
+    (*new_accumulator)->rank_cumul_rw_only_time = &((*new_accumulator)->rank_cumul_io_total_time[job_nprocs]);
+    (*new_accumulator)->rank_cumul_md_only_time = &((*new_accumulator)->rank_cumul_io_total_time[job_nprocs*2]);
+
+    return(0);
+}
+
+int darshan_accumulator_inject(darshan_accumulator acc,
+                               void*               record_array,
+                               int                 record_count)
+{
+    int i;
+    void* new_record = record_array;
+    uint64_t rec_id;
+    int64_t r_bytes;
+    int64_t w_bytes;
+    int64_t max_offset;
+    int64_t nprocs;
+    int64_t rank;
+    double io_total_time;
+    double md_only_time;
+    double rw_only_time;
+    int ret;
+    file_hash_entry_t *hfile = NULL;
+
+    if(!mod_logutils[acc->module_id]->log_agg_records ||
+       !mod_logutils[acc->module_id]->log_sizeof_record ||
+       !mod_logutils[acc->module_id]->log_record_metrics) {
+        /* this module doesn't support this operation */
+        return(-1);
+    }
+
+    for(i=0; i<record_count; i++) {
+        /* accumulate aggregate record */
+        if(acc->num_records == 0)
+            mod_logutils[acc->module_id]->log_agg_records(new_record, acc->agg_record, 1);
+        else
+            mod_logutils[acc->module_id]->log_agg_records(new_record, acc->agg_record, 0);
+        acc->num_records++;
+
+        /* retrieve generic metrics from record */
+        ret = mod_logutils[acc->module_id]->log_record_metrics( new_record,
+            &rec_id, &r_bytes, &w_bytes, &max_offset, &io_total_time,
+            &md_only_time, &rw_only_time, &rank, &nprocs);
```

Comment:
This was kind of a vague nod to future proofing in case one of the modules gains the ability to report partial reductions (e.g. partial sharing) in a future update, either by using a better runtime reduction algorithm or by post-processing.  TBF, though, none of them can actually do that right now, so it's not really serving any real purpose at the moment.

I could go either way; it won't change functionality to cut it and just use the rank.  What do you think?

Reply:
Oh, yeah, I think that's a good idea to keep it for that reason actually. Not a big deal really, just something I noticed when reading through.